src/raw_node.rs

```diff
@@ -349,94 +379,259 @@ impl<T: Storage> RawNode<T> {
-            // whether they were included in rd.HardState or not.
-            self.advance_apply(commit_idx);
+    /// Notifies that the ready of this number has been well persisted.
+    ///
+    /// Since Ready must be persisted in order, calling this function implicitly means
+    /// all readys with numbers smaller than this one have been persisted.
+    pub fn on_persist_ready(&mut self, number: u64) {
+        let (mut index, mut term) = (0, 0);
+        while let Some(record) = self.records.front() {
+            if record.number > number {
+                break;
+            }
+            let mut record = self.records.pop_front().unwrap();
+
+            if let Some((i, t)) = record.last_entry {
+                index = i;
+                term = t;
+            }
+
+            if !record.messages.is_empty() {
+                self.messages.push(mem::take(&mut record.messages));
+            }
         }
+        if term != 0 {
+            self.raft.on_persist_entries(index, term);
+        }
+    }
+
+    /// Advance notifies the RawNode that the application has applied and saved progress
+    /// in the last Ready results.
+    ///
+    /// Returns the LightReady that contains commit index, committed entries and messages.
+    pub fn advance(&mut self, rd: Ready) -> LightReady {
+        let applied = self.commit_since_index;
+        let light_rd = self.advance_append(rd);
+        self.advance_apply_to(applied);
+        light_rd
+    }
+
+    /// Advance append the ready value synchronously.
+    ///
+    /// Returns the LightReady that contains commit index, committed entries and messages.
+    ///
+    /// Since Ready must be persisted in order, calling this function implicitly means
+    /// all readys collected before have been persisted.
+    #[inline]
+    pub fn advance_append(&mut self, rd: Ready) -> LightReady {
+        self.commit_ready(rd);
+        self.on_persist_ready(self.max_number);
+        let mut light_rd = self.gen_light_ready();
+        // Set commit index if it's updated
+        let hard_state = self.raft.hard_state();
+        if hard_state.commit > self.prev_hs.commit {
+            light_rd.commit_index = Some(hard_state.commit);
+            self.prev_hs.commit = hard_state.commit;
+        } else {
+            assert!(hard_state.commit == self.prev_hs.commit);
+            light_rd.commit_index = None;
+        }
+        assert_eq!(
+            hard_state, self.prev_hs,
+            "hard state {:?} != prev_hs {:?}",
+            hard_state, self.prev_hs
+        );
+        light_rd
     }
 
-    /// Appends and commits the ready value.
+    /// Advance append the ready value asynchronously.
     #[inline]
-    pub fn advance_append(&mut self, rd: Ready) {
+    pub fn advance_append_async(&mut self, rd: Ready) {
         self.commit_ready(rd);
     }
 
+    /// Advance apply to the index of the last committed entries given before.
+    #[inline]
+    pub fn advance_apply(&mut self) {
```

Comment:
Why not just keeping the original interface and implementations?

Reply:
If so, what's another function's name which uses `commit_since_index` for sync applying?

Reply:
What I mean is just keep one function that accepts explicit index is enough. Application always knows what index has been applied.

Reply:
The function which uses `commit_since_index` is convenient for users. For the previous implementation, users don't need to record the applied index if the logs are applied synchronously. As our examples, this function is used after handling the `LightReady`.